src/parsec_service/stability.md

```diff
@@ -0,0 +1,68 @@
+# Parsec Stability
+
+Work in progress, see the [related issue](https://github.com/parallaxsecond/parsec-book/issues/83).
+
+## Why is stability important?
+
+Parsec is a long-running service that will be configured, deployed, upgraded and maintained over a
+long period of time. The environment will also have client applications that are calling it, and
+persistent state that is collected on storage. Components will be upgraded over time. For Parsec to
+be supportable in enterprise system, it must not be brittle against upgrades. If Parsec is upgraded
+and restarted, client applications should continue working, and all persistent state from prior
+versions should remain valid and usable.
+
+## What does stability mean for the Parsec service?
+
+The following definition of stability is chosen for the Parsec service: *whatever two versions A and
+B of the Parsec binary, B being newer than A, A and B are stable if A can be removed and replaced by
+B without breaking anything*.
+
+Note that we are not looking at stability in the reverse direction: B can not be replaced by A
+without breaking anything.
+
+The principle of [semantic versioning](https://semver.org/) is used to describe the stability of
+Parsec versions and when breaking changes are done. If `parsec` version is at `1.0.0` then all
+future stable version to that will be `1.x.y`.
+
+## What needs to be stable?
+
+With the above definition, Parsec stays stable if it keeps the exact same communication channels and
+format with its environment. Those are already described in the [dataflow
+diagram](../parsec_security/parsec_threat_model/threat_model.md#dataflow-diagram) in the threat
+model.
+
+The communication channels that needs to be stable are (some of them are not in the diagram):
+
+1. Communication with clients
+2. Communication with authenticators
```

Comment:
Does this include the paths used to connect clients to service or service to authenticators?

Reply:
Also, do they cover the "logical" contents of the communication? For example, if one version of the service (or of a provider) accepts some set of parameters and another doesn't (or the other way around), is that a breaking change? Fixing a bug shouldn't really fall into "breaking changes", but how big of a bug can we allow that for?

The list keys bug is a good example (though a bit contrieved): in one version one key reports as belonging to `MbedCryptoProvider`, in the next as `TpmProvider` - is that a breaking change? I guess this isn't that controversial because the first one is clearly wrong, but I can imagine cases where something might be either treated as an error or just ignored, and we switch from one to the other.

Reply:
> Does this include the paths used to connect clients to service or service to authenticators?

1 is for clients to service and 2 for service to authenticators.

> Also, do they cover the "logical" contents of the communication? For example, if one version of the service (or of a provider) accepts some set of parameters and another doesn't (or the other way around), is that a breaking change? Fixing a bug shouldn't really fall into "breaking changes", but how big of a bug can we allow that for?
>
> The list keys bug is a good example (though a bit contrieved): in one version one key reports as belonging to MbedCryptoProvider, in the next as TpmProvider - is that a breaking change? I guess this isn't that controversial because the first one is clearly wrong, but I can imagine cases where something might be either treated as an error or just ignored, and we switch from one to the other.

That's a good question. My first feeling would that this only covers the format of the communication: how the messages are sent and received but not their logical content. I would say that for the logical content clients can make stable runtime decisions without breaking if the content changes. For example we could not publish a new breaking change each time "ListOpcodes" returns one more opcode but clients can make decisions based on that that will be stable in the future.

Maybe the definition of stable ("without breaking anything") is a bit too loose as obviously a lot of things could break if `ListKeys` would only return "MbedCryptoProvider".